Add docs on how to think through inline directives (#22)
aristanetworks/purescript-backend-optimizer: docs/why-and-how-of-inline-directives.md

someComputation :: forall a. Int -> FailureOrSuccess a

-- Module2.purs
resultToEither :: FailureOrSuccess a -> Either String a
resultToEither arg = case  of
  Failure -> Left "There was an error!"
  Success a -> Right a

computationAsEither = resultToEither $ someComputation 4

-- Module3.purs
eitherToMaybe :: Either String a -> Maybe a
eitherToMaybe arg = case arg of
  Left _ -> Nothing
  Right a -> Just a

computationAsMaybe = map doSomethingElse computationAsEither

-- Module4.purs
maybeToBoolean :: Maybe a -> Boolean
maybeToBoolean arg = case arg of
  Nothing -> false
  Just a -> true

-- Module5.purs
foo = do
  let x = maybeToBoolean computationAsMaybe
  ...
```

</details>

At the end of the day, evaluating `case1` and `case2` produce the same boolean value, but `case1` produces and then consumes 2 intermediate data structures unnecessarily: `Either` and `Maybe`.

**To summarize, the goal of our optimizations is to remove these unneeded intermediate data structures.** As a result, the code is more performant because there's less purely-overhead work for the computer to do at runtime.

## How Code Gets Optimized

### Primitive Data Flow: Production and Consumption

> How does `purs-backend-es` know when unneeded intermediate data structures are being used and when it is safe to remove them? These usages arise when primitive data is "produced" and then immediately "consumed".

Reviewer: I'm not sure the producer/consumer nomenclature is quite appropriate here. More commonly you'll see "constructor" and "eliminator". It's the same duality though. That is, `Right` is the constructor, and `case` is the eliminator.

Reviewer: An accessor being sugar for a primitive `case`.

```
getFoo = case _ of
  { foo } -> foo
```

eitherToMaybe $ Right "value"
```

Here's another example using records. The `{ bar: 42 }` record is "produced" by being defined in a let binding. It is then immediately "consumed" when its `bar` field is accessed. Thus, the record can be removed entirely and `a.bar` can be replaced with `42`.

```purs
foo = do
  let
    a = { bar: 42 }
    b = 1 + a.bar
  b
```

Lastly, here's an example with a few different conclusions:

- If we do NOT know what `f` will do with `a` (e.g. if `f` was some FFI function), then we cannot call the record an "unnecessary intermediate data structure." `f` may need the entire record, and that possibility forces us to call the record structure necessary.
- If we do know what `f` will do (e.g. if `f = _.bar`), then this is still an example of a primitive data "production" followed by immediate primitive data "consumption".

```purs
foo = do
  let
    a = { bar: 42, baz: "something" }
    b = 1 + f a
  b
```

**To summarize, a data producer followed immediately by a data consumer indicates a place where an unneeded intermediate data structure exists and can be removed.**

### Removing Unneeded Data Structures

#### Inlining and Inline Directives

> An inliner replaces a function's call site (e.g. stuff on the left-hand-side of the `=`) with its implementation (e.g. stuff on the right-hand-side of the `=`). This replacement means the function's body is duplicated and will appear at least twice in the resulting code: once in its original definition and once in the usage site.

Reviewer: This is a nit, but the left-hand-side of `=` is not a call-site. I think you can just say replaces a "function call", and then it would be more accurate.

Here's one example using the function `binaryPlus`:

```purs
binaryPlus :: Int -> Int -> Int
binaryPlus a b = a + b

usage = 9 + binaryPlus 1 2
```

There's a few ways we could inline the expression, `binaryPlus 1 2`. While `purs-backend-es` will performance its own analysis to determine whether and when to inline it, it doesn't know as much as the developer does.

A developer can use **inline directives** to tell `purs-backend-es` whether to inline a function's body to its usage site and when (i.e. how many args need to be passed to the function before the inlining occurs):

- never inline it at all (i.e. the directive is `never`)
    - `9 + binaryPlus 1 2`
- always inline it with no consideration for the number of arguments applied to it (i.e. the directive is `always`)
    - `9 + (\a b -> a + b) 1 2`
- inline it only after at least one argument has been applied to it (i.e. the directive is `arity=1`)
    - `9 + (\  b -> 1 + b)   2`
- inline it only after at least two arguments have been applied to it (i.e. the directive is `arity=2`)

>     - `9 + (        1 + 2)`

Reviewer: I think this example is a little confusing because there's no difference here in practice between the arities here since they call is saturated anyway. To illustrate that you have to show some partial application. That is, with `arity=2` something like `binaryPlus 1` would not inline. However with `arity=1`, you'd get `\b -> 1 + b`.

**To summarize, inline directives provide developers with a stronger guarantee about whether and when a function is inlined than just relying upon the analysis done by `purs-backend-es`.**

#### Inlining Duplicates Code

The below example highlights the risk of inlining everything without thought: pointless code duplication.

```purs
hasExpensiveLetBinding a b = do
  let
    someValue = expensiveComputation
  1 + someValue * b + a

usage1 = hasExpensiveLetBinding 1 2
usage2 = hasExpensiveLetBinding 3 4
```

If we inline `hasExpensiveLetBinding` immediately, the `expensiveComputation` will be computed three times rather than once. In other words, it would be the same as writing the following in source code:

```purs
hasExpensiveLetBinding a b = do
  let
    someValue = expensiveComputation
  1 + someValue * b + a

usage1 = (\a b -> do
    let
      someValue = expensiveComputation
    1 + someValue * b + a
  ) 1 2
usage2 = (\a b -> do
    let
      someValue = expensiveComputation
    1 + someValue * b + a
  ) 3 4

> ```

Reviewer: This example is inaccurate. Inlining here has no effect on how many times `expensiveComputation` is computed, it's always 2 because `expensiveComputation` is deferred under the lambda application.

Author: What would be an example that demonstrates this effect while still being correct?

Reviewer: Maybe you should just reframe it not in terms of how expensive it is to compute/evaluate, but rather in terms of code size. That is, if `expensiveComputation` (should be renamed) is a large expression that won't be reduced by the optimizer based on usages of `a` or `b`, then you've duplicated code without also reducing runtime overhead.

**To summarize, inlining for the sake of inlining is a great way to unnecessarily increase your program's bundle size**.

> #### Rewrite Rules

Reviewer: While the name of this section is not inaccurate, it's really confusing if you are familiar with GHC rewrite rules. "Rewrite rules" don't appear anywhere in the source code. So, yes, technically there are rules for how the optimizer rewrites things, but I would steer very clear of calling it that.

Reviewer: Instead I would probably just call this evaluation. The optimizer knows how to evaluate PureScript code, and will evaluate things at build time.

Author: > I would steer very clear of calling it that.

Ok, I'll do that.

> will evaluate things at build time.

This isn't quite correct either as there are still some things it won't evaluate. For example:
```purs
mapWithIndex Tuple [ 1, 2, 3]
```
doesn't get evaluated to `[ Tuple 0 1, Tuple 1 2, Tuple 2 3 ]` because no such "rule" exists. However, calling it something else like "evaluation rules", which may be more accurate, is too similar to "rewrite rules".

Reviewer: All I'm saying is that we shouldn't invent terminology that doesn't exist in the project's code base, and doesn't co-opt terminology used in another similar, high-profile project. Evaluation is what's happening, which is why the main function is called `eval`.

Author: How about this rendering?

> `purs-backend-es` _can_ evaluate PureScript code, but it will only evaluate _some_ expressions at build time. Typically, such expressions involve literal values (e.g. `1`, `{ foo: "string" }`). For example, primitive addition can be evaluated when its arguments are both literal `Int` values (e.g. `1 + 2` is evaluated to `3`). The below code demonstrates this:

A **rewrite rule** will replace one expression, usually involving literal values, with an equivalent expression. For example, an expression like `1 + 2` will always produce the value `3`. So, we can replace that expression with `3`. The rewrite rule here could be defined as, "Whenever the `+` function is called on two literal `Int` values, replace that expression with the sum of the two `Int` values."

```purs
-- Before
foo = 1 + 2

>   --  ^^^^^ Rewrite rule: "I can optimize that! `1 + 2` is `3`"

Reviewer: Ie, the optimizer knows how to evaluate primitive addition, and it sees that `1 + 2` can be reduced to `3`.

-- After
foo = 3
```

However, some expressions will prevent such rewrite rules from triggering. For example:

```purs
foo =
  let
    a = { bar: 42 }
  in
    1 + a.bar
--  ^^^^^^^^^ Rewrite rule: "I can't optimize that! `a.bar` is not a literal `Int` value."
```

Because the rewrite rule cannot see the literal `Int` value represented by `a.bar`, its optimization does not trigger. To solve this problem, we need to make it visible. The solution is to duplicate code via inlining.

> **To summarize, inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes.**

Reviewer: ```suggestion
**To summarize, inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise not apparent.**
```
👀

#### Code Optimzation Example via Rewrite Rules and Inlining

Let's see how this works in practice using this example:

```purs
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2

foo = ignoreArgs 8 9

> ```

Reviewer: This whole series of steps is not quite accurate because the optimizer will optimizer ignoreArgs on it's own to
```
ignoreArgs arg1 arg2 =
  42 + arg1 + arg2
```
Before it even does anything with `foo`.

First, we see that `ignoreArgs` has 2 arguments applied. Since `2` >= `1`, the arity of the directive we specified, we inline `ignoreArgs`. Since `8` corresponds to `arg1`, we replace `arg1` with `8` when inlining `ignoreArgs`. We don't remove the original `ignoreArgs` declaration.

```purs
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2

foo = (\arg2 -> do
  let
    a = { bar: 41 }
  1 + a.bar + 8 + arg2
  ) 9
```

At this point, `purs-backend-es`' default inliners will see a lambda (i.e. a "producer") being immediately applied to an argument (i.e. a "consumer"). Thus, it will inline that argument into the lambda's body. This gets us:

```purs
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2

foo = do
  let
    a = { bar: 41 }
  1 + a.bar + 8 + 9
```

A rewrite rule will detect the `8 + 9` expression, and replace it with its sum: `17`. However, `1 + a.bar` and `a.bar + 17` don't trigger the rewrite rule because one of the values is not a literal `Int` value.

```purs
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2

foo = do
  let
    a = { bar: 41 }

>   1 + a.bar + 17

Reviewer: This example is somewhat confusing because `+` is commutative. That is, if the optimizer was smart enough it could see that it's equivalent to `a.bar + 18`. In reality, it only currently tries to evaluate around abstract terms associatively, and even then only to a fixed depth, because doing this in general is quadratic.

So the example is accurate, but it invites more questions from the reader that I don't think you want to answer here.

```

Fortunately, `purs-backend-es`' default inliners will see a record binding (i.e. a "producer") that is immediately accessed under the label `bar`. Thus, the corresponding value will be inlined:

```purs
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2

foo = do
  let
    a = { bar: 41 }
  1 + 41 + 17

> ```

Reviewer: This is inaccurate. The optimizer uses analysis to determine whether or not a binding should be inlined, and it will do that in lieu of creating a binding. So in this case, for `ignoreArgs`, what it actually does is: 
```
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2
```
`a` is a record constructor and is eliminated in all references, so it can be inlined.
```
ignoreArgs arg1 arg2 = do
  1 + { bar: 42 }.bar + arg1 + arg2
```
The record constructor is immediately eliminated, so it can be reduced:
```
ignoreArgs arg1 arg2 = do
  1 + 42 + arg1 + arg2
```
The operands of + are both known literals (up to some fixed depth, associatively), so it can be reduced.
```
ignoreArgs arg1 arg2 = do
  43 + arg1 + arg2
```

    a = { bar: 41 }
  59
```

Lastly, `purs-backend-es` determines that the `a` binding is never used; thus, it's removed entirely:

```purs
ignoreArgs arg1 arg2 = do
  let
    a = { bar: 41 }
  1 + a.bar + arg1 + arg2

foo = 59
```

Now that we've finished optimizing `foo`, the following states are true:

- `ignoreArgs` is still defined as it originally was
- the optimized `foo` still represents the same value as the unoptimized `foo` would have had at runtime.

## A Methodology for Defining Inline Directives

Inlining duplicates code so that rewrite rules can trigger optimizations that were otherwise hidden from its eyes. Ideally, inlining values will always trigger rewrite rules that both reduce the bundle size of the code AND make the resulting program more performant. However, inlining may increase a program's bundle size without improving its performance.

In short, this process isn't scientific. While one can add inline directives without much thought, the result likely won't be what they want.

So, how should one determine if an inline directive needs to be added? One should always use the below methodology:

1. Think of a snippet of code you want to ensure is optimized.

> 2. Define a snapshot for that snippet.

Reviewer: What is a snapshot and how does the reader define one? There's a lot of implicit knowledge here that makes no sense to most readers.

Reviewer: Is the audience contributors to the project, or users of the backend?

Author: The intended audience are users of the backend. I guess rather than 'snapshot', I should continue using 'snippet'. However, an example that clarifies what a good snippet would look like may help illustrate/define this term more.

3. Look at the current JavaScript output of that snapshot.
4. If the current output is not good enough,
    1. add an inline directive to the outermost thing
       1. When in doubt, use `arity=x` where `x` is the number of args the function in the outputted JavaScript, not in PureScript source code, takes.
    2. Go to Step 3.
5. Once the output satisfies you, determine where it should be defined (see next section)

First, think of a snippet of code you want to ensure is optimized. If you don't have a goal in mind, you will add inline directives that will unnecessarily bloat your code.

Second, define a snapshot for that snippet. Without a small snippet of code, you won't be able to see what affects adding more directives may have.

Third, look at the current JavaScript output of that snapshot. If it's already as optimized, there's no reason to add an inline directive. You're done.

> Otherwise, fourth, find the outermost function and add an inline directive for that. Because `purs-backend-es` operates on `CoreFn`, not PureScript source code, refer to the function in the outputted JavaScript to determine how many args the function takes.

Reviewer: Is this just repeating the steps above?

Author: Yes, but each expands on the heading and provides more context, except... I thought there would be more to say for each step, but apparently not.
